zmq.hpp

```diff
@@ -1082,36 +1082,46 @@ namespace zmq
                 throw error_t ();
         }
 
-        bool wait (std::chrono::milliseconds timeout)
+        int wait (std::chrono::milliseconds timeout)
         {
-            int rc = zmq_poller_wait_all (poller_ptr, poller_events.data (), static_cast<int> (poller_events.size ()), static_cast<long>(timeout.count ()));
-            if (rc >= 0) {
-                std::for_each (poller_events.begin (), poller_events.begin () + rc, [](zmq_poller_event_t& event) {
+            if (need_rebuild) {
+                poller_events.clear ();
+                poller_handlers.clear ();
+                poller_events.reserve (handlers.size ());
+                poller_handlers.reserve (handlers.size ());
+                for (const auto &handler : handlers) {
+                    poller_events.emplace_back (zmq_poller_event_t {});
+                    poller_handlers.push_back (handler.second);
+                }
+                need_rebuild = false;
+            }
+            int rc = zmq_poller_wait_all (poller_ptr, poller_events.data (),
+                                          static_cast<int> (poller_events.size ()),
+                                          static_cast<long>(timeout.count ()));
+            if (rc > 0) {
+                std::for_each (poller_events.begin (), poller_events.begin () + rc,
+                               [](zmq_poller_event_t& event) {
                     if (event.user_data != NULL)
                         (*reinterpret_cast<handler_t*> (event.user_data)) (event.events);
                 });
-                return true;
+                return rc;
             }
-
 #if ZMQ_VERSION >= ZMQ_MAKE_VERSION(4, 2, 3)
             if (zmq_errno () == EAGAIN)
 #else
             if (zmq_errno () == ETIMEDOUT)
 #endif
-                return false;
+                return 0;
 
             throw error_t ();
         }
 
-        size_t size ()
```

Comment:
I don't get why removing this method is necessary. There are several other options: get the size before pending changes, get the size after pending changes, provide getters for both.

Reply:
what value does it bring to the user? Would it not be more confusing than useful?
poller is not really a collection, you can not iterate through it etc. and user needs to store info about registered socket already in order to successfully unregistered it later.

Reply:
Maybe the size is not a relevant information, but the information if the poller is empty or not is, since this affects how wait may be called the next time. Therefore, this information should take into account that pending changes are applied. (I am not sure how you implemented handling of pending changes, but I think that there can only be pending changes if empty is called during an active call of wait, right? Actually, such reentrant calls are not relevant in the use cases I had in mind previously).

Overall I would be fine with removing the size method if an empty method is added instead.

Reply:
..but `empty` has the same problem as `size`. For example:
```
zmq::poller_t poller;
poller.add (socket, ZMQ_POLLIN, [&](short events) {
    poller.remove (socket);
    bool empty = poller.empty();
    // hmmm
});
```
 What would you expect `empty` to be here? You are in the poller's event processing loop but just has removed last socket. If you still insist on adding `empty` or `size` I would go for `true` in the example above as it reflects latest user's actions.

Reply:
just created a PR for that.